finat/finiteelementbase.py

```diff
@@ -32,6 +32,31 @@ def entity_dofs(self):
         '''Return the map of topological entities to degrees of
         freedom for the finite element.'''
 
+    @property
+    def entity_permutations(self):
+        '''Returns a nested dictionary that gives, for each dimension,
+        for each entity, and for each possible entity orientation, the
+        DoF permutation array that maps the entity local DoF ordering
+        to the canonical global DoF ordering.
+
+        The entity permutations `dict` for the degree 4 Lagrange finite
+        element on the interval, for instance, is given by:
+
+        .. code-block:: python3
+
+            {0: {0: {0: [0]},
+                 1: {0: [0]}},
+             1: {0: {0: [0, 1, 2],
+                     1: [2, 1, 0]}}}
```

Comment:
Are the keys of the innermost dict always just `range(num_orientations)` ? If so, do we need keys?

Reply:
If I understand it, this is effectively memoizing the action of each group member for every entity, knowing that the group is a permutation group. Is this right?

Reply:
Yes, you are right.
In my current notation, the keys of the inner most dict are indeed always `range(num_orientations)`, and I should have just used `tuple`s instead of `dict`s. If we wanted to adopt PETSc style here, though, would we have something like:
```
{0: {0: {0: [0]},
     1: {0: [0]}},
1: {0: {0: [0, 1, 2],
        -1: [2, 1, 0]}}}
```
as per Table 1.3 of Matt's note?

Reply:
As I noted here https://github.com/firedrakeproject/firedrake/pull/2101, using PETSc notation for orientations here would be ideal, but seems not straightforward until we change how we construct `cell_closure`. I wonder if what I'm saying makes sense.